### DISCUSSION TITLE
Interpolated String Parse-Error

### DISCUSSION THREAD

**Discussion by bernd5** [dotnet/csharplang#4361] [Language Ideas]
**Version Used**: 
3.9.0-2.final

**Steps to Reproduce**:
The following code is parsed invalid:
```CSharp
using System;

namespace Tests
{
	class App
	{
		static void Main()
		{
			string s = $"InterpolatedString {{{27,7:X}}}";

			Console.WriteLine(s);
		}
	}
}
```

Easy visible with syntax highlighting:
![grafik](https://user-images.githubusercontent.com/26689688/102718120-c16d2900-42e6-11eb-8fcd-cae26ab1e864.png)

**Expected Behavior**:
The curly braces do not belong to InterpolationFormatClauseSyntax. The content should be simply: "X".

**Actual Behavior**:
we get an InterpolationFormatClauseSyntax with the content "X}"

**Comment by svick** [comment]
While this behavior is definitely confusing, it's consistent with [what `string.Format` does in the same situation](https://docs.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting?WT.mc_id=DOP-MVP-5002735#escaping-braces), so I think this is the expected behavior.

**Comment by bernd5** [comment]
Is there any meaningfull format-string which needs curly-braces?

**Comment by bernd5** [comment]
The documentation you posted seems to be outdated. The current implementation does not print "{D}" (see [here](https://dotnetfiddle.net/okys5Q)).

It works just as expected. Sample code:
```CSharp
Console.WriteLine($"{{{12:X}}}"); //prints "{C}"
````

which is lowered to:
```CSharp
Console.WriteLine(string.Format("{{{0:X}}}", 12));
```

And the final string is "{C}".

**Comment by bernd5** [comment]
I looked up the [implementation](https://github.com/dotnet/runtime/blob/c72678bac90f9692ccea2787e22d88a6ae449cda/src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs#L1701):

- the first '}' terminates the format-string-part
- '{' in the format-string would cause a "FormatError"

**Comment by PathogenDavid** [comment]
> While this behavior is definitely confusing, it's consistent with [what `string.Format` does in the same situation](https://docs.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting?WT.mc_id=DOP-MVP-5002735#escaping-braces), so I think this is the expected behavior.

I dug around a little, and it turns out that documentation is out of date. ([SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEBMAjALABQZuhADAARQ0C8NAJAEQAScANlwPY0BvIYXxIA7CAAaAX1msA3BUIBOABRQAlItKU1bIQJFTZ01lrJkAlphg0AbhC7o4jGgDYAzPgAs2yrV50GAAHINd/ADoAMV4ESBhVVgMqEAARE1Yke0dncx0VVUCQoK0gA==))

The logic was changed to align with what people generally expect in .NET Core 3.0 in https://github.com/dotnet/coreclr/pull/23062. (Oddly enough it seems this wasn't reported on the list of [breaking changes in .NET Core 3.0](https://docs.microsoft.com/en-us/dotnet/core/compatibility/3.0).)

It seems the Roslyn team never got the memo. It doesn't actually matter for the actual lowering because the format string isn't any different in the end, but the syntax tree represents things incorrectly.

**Comment by svick** [comment]
@bernd5 @PathogenDavid I didn't know about that, thanks for teaching me something new. I have opened https://github.com/dotnet/docs/issues/22156 to make sure the documentation is updated.

**Comment by jaredpar** [comment]
> It seems the Roslyn team never got the memo.

Correct for me at least. 😄

> It doesn't actually matter for the actual lowering because the format string isn't any different in the end, but the syntax tree represents things incorrectly.

The "incorrectly" part is a bit dicey here. Consider that the compiler has to compile for a variety of runtimes here: some of which where the behavior is present and some where it is not. Hence in some cases it's correct, and some cases it's not :smile:

I'm not sure there is really a lot we can do here at the syntax level. Possible that we could decide that say C# 8 and above we parse interpolated strings differently (given that C# 8 is associated with the runtime version where this change occurred). Not sure if that is worth the change or not. 

@CyrusNajmabadi have any thoughts here?

**Comment by bernd5** [comment]
In my eyes the old behaviour is never useful. It is just a "documented bug" in String.Format.
It would be useful only if there were a need for curly braces in format strings which are not expressable by the new syntax.

If we would change the parsing rules - the runtime behaviour of old runtimes would not change as long as the lowering is untouched. Therefore I don't see a need for language-version specific parsing.

**Comment by CyrusNajmabadi** [comment]
Given that the runtime took the breaking change, I'm totally ok with the language doing the same. I agree that it's not really realistic that anyone would have every wanted or depended on the previous behavior.

**Comment by jaredpar** [comment]
Actually upon thinking this is more of a lanuage issue vs. compiler hence moving to csharplang

**Comment by bernd5** [comment]
I had a look at the [spec](https://github.com/dotnet/csharplang/blob/a3d67d141501101af9dbb9e8192ca8b4c717e316/spec/lexical-structure.md#interpolated-string-literals) and found:

```
interpolation_format_character
    : '<Any character except \" (U+0022), : (U+003A), { (U+007B) and } (U+007D)>'
    ;
```

Therefore the roslyn implementation seems to be not spec-conformant. The issue can move back to roslyn.

**Comment by jaredpar** [comment]
@bernd5 the issue though is should we adjust to match the new interpretation of the BCL.

**Comment by alrz** [comment]
while you're at it, maybe permit ternaries without parentheses in case this is actually a parsing issue.

I think Cyrus had a proposal to parse it just like how TS does.

**Comment by bernd5** [comment]
As I understand the spec, the language forces the new interpretation of BCL for interpolation-format-strings.
And I understand that this was not a big issue because the current roslyn implementation allows both interpretations by simply not interpreting it at compile-time. It is handled at runtime via String.Format.

We get just an issue if we write for example syntax-rewriters to optimize such strings. And of course things like syntax-highlighting.

**Comment by bernd5** [comment]
In addition to the "brace-issue" I would like to add these issues, too:

https://github.com/dotnet/roslyn/issues/50742

https://github.com/dotnet/roslyn/issues/50741

**Reply by CyrusNajmabadi** [reply]
These are out of scope for this discussion.

**Comment by bernd5** [comment]
In addition the spec does not dissallow preprocessor directives / the hash-sign inside of interpolated strings which are explicitly dissallowed in the roslyn implementation (who is right?).

Shouldn't preprocessing handled before lexing?

This code does not compile with roslyn:
```
using System;

Console.WriteLine(@$"Test
{
#if XXX
  "A" 
#else
  "B"
#endif 
}");
```

**Reply by CyrusNajmabadi** [reply]
> which are explicitly dissallowed in the roslyn implementation (who is right?).

Roslyn both follows the spec and is correct here in the impl.

> Shouldn't preprocessing handled before lexing?

No.  That is not how the language is spec'ed.  handling of preprocessor directives happens during the lexical pass.

**Reply by bernd5** [reply]
In chsarp we have (luckily) just limited preprocessing support (no macros ...) inherited from C.
In general preprocessing directives should behave if they were preprocessed - but they are not really preprocessed mostly because of performance reasons.

And there is a single exception: verbatim string literals ([spec](https://github.com/dotnet/csharplang/blob/a3d67d141501101af9dbb9e8192ca8b4c717e316/spec/lexical-structure.md#conditional-compilation-directives)).

In my eyes it is okay that preprocessing in not supported in expression-holes, but it should be explicitly mentioned in the spec.

**Reply by CyrusNajmabadi** [reply]
> In general preprocessing directives should behave if they were preprocessed 

They are processed by the lexer. This is intentional and has been the design of c# since 1.0. This will not change as it's writing as intended and would literally break code of we changed the meaning here.

**Comment by bernd5** [comment]
By updating to net6.0 the interpolated string result becomes buggy again. 

Now:
```cs
Console.WriteLine($"{{{12:X}}}");
```

which was lowered to:


```cs
Console.WriteLine(string.Format("{{{0:X}}}", 12)); //prints "{C}"
```

is lowered to

```cs
DefaultInterpolatedStringHandler defaultInterpolatedStringHandler = new DefaultInterpolatedStringHandler(1, 1);
defaultInterpolatedStringHandler.AppendLiteral("{");
defaultInterpolatedStringHandler.AppendFormatted(12, "X}}");
Console.WriteLine(defaultInterpolatedStringHandler.ToStringAndClear()); //prints "{X}}"
```

instead of:

```cs
DefaultInterpolatedStringHandler defaultInterpolatedStringHandler = new DefaultInterpolatedStringHandler(2, 1);
defaultInterpolatedStringHandler.AppendLiteral("{");
defaultInterpolatedStringHandler.AppendFormatted(12, "X");
defaultInterpolatedStringHandler.AppendLiteral("}");
Console.WriteLine(defaultInterpolatedStringHandler.ToStringAndClear()); //prints "{C}"
```

Could we please fix the parsing?

**Reply by CyrusNajmabadi** [reply]
@bernd5 this has already been fixed. In the future, please file issues at roslyn, not csharplang.

**Reply by bernd5** [reply]
@CyrusNajmabadi great to hear (which version?)! This discussion was moved from roslyn...

**Comment by CyrusNajmabadi** [comment]
I'd be willing to champion a change here. The runtime already decided to take a break here and they believe the new behavior is more desirable.  Given that upgrading the corresponding runtime breaks you here if you depended on it, then I don't see why it would be considered worse if the language did he same. 

Thoughts @jaredpar ?

I've been pretty deep into this code, and it wouldn't be a problem to do.
